src/hazelweb/gui/CursorInspector.re

```diff
@@ -5,21 +5,516 @@ type err_state_b =
+        inconsistent_symbol,
+        emphasize_text(string_of_int(got_len) ++ "-tuple"),
+      ]
+    | AnaInvalid(expected_ty)
+    | PatAnaInvalid(expected_ty) => [
+        ana,
+        HTypCode.view(expected_ty),
+        inconsistent_symbol,
+        emphasize_text("Invalid Text"),
+      ]
+    | SynInvalid => [emphasize_text("Invalid Text")]
+    | SynInvalidArrow(_) => [
+        syn,
+        emphasize_text("Function Type"),
+        inconsistent_symbol,
+        emphasize_text("Invalid Text"),
+      ]
+    | AnaFree(expected_ty) => [
+        ana,
+        HTypCode.view(expected_ty),
+        inconsistent_symbol,
+        emphasize_text("Free Variable"),
+      ]
+    | SynFree => [syn, emphasize_text("Free Variable")]
+    | SynFreeArrow(_) => [
+        syn,
+        emphasize_text("Function Type"),
+        inconsistent_symbol,
+        emphasize_text("Free Variable"),
+      ]
+    | AnaKeyword(expected_ty, keyword) =>
+      let main_msg = [
+        ana,
+        HTypCode.view(expected_ty),
+        inconsistent_symbol,
+        emphasize_text("Reserved Keyword"),
+      ];
+      keyword_msg(term, keyword, main_msg);
+    | PatAnaKeyword(expected_ty, _) => [
+        ana,
+        HTypCode.view(expected_ty),
+        inconsistent_symbol,
+        emphasize_text("Reserved Keyword"),
+      ]
+    | SynKeyword(keyword) =>
+      let main_msg = [syn, emphasize_text("Reserved Keyword")];
+      keyword_msg(term, keyword, main_msg);
+    | PatSynKeyword(_) => [syn, emphasize_text("Reserved Keyword")]
+    | SynKeywordArrow(_, keyword) =>
+      let main_msg = [
+        syn,
+        emphasize_text("Function Type"),
+        inconsistent_symbol,
+        emphasize_text("Reserved Keyword"),
+      ];
+      keyword_msg(term, keyword, main_msg);
+    | SynBranchClause(join, typed, _) =>
+      switch (join, typed) {
+      | (JoinTy(ty), Synthesized(got_ty)) =>
+        if (HTyp.consistent(ty, got_ty)) {
+          [syn, HTypCode.view(ty)];
+        } else {
+          let (ty_diff, got_diff) = TypDiff.mk(ty, got_ty);
+          [
+            syn,
+            HTypCode.view(~diff_steps=ty_diff, ty),
+            inconsistent_symbol,
+            HTypCode.view(~diff_steps=got_diff, got_ty),
+          ];
+        }
+      | _ => message(typed)
+      }
+    | SynInconsistentBranches(_) => [
+        syn,
+        emphasize_text("Inconsistent Branch Types"),
+      ]
+    | SynInconsistentBranchesArrow(_) => [
+        syn,
+        emphasize_text("Function Type"),
+        inconsistent_symbol,
+        emphasize_text("Inconsistent Branch Types"),
+      ]
+    | OnType => []
+    | OnNonLetLine => /* TODO */ [emphasize_text("Line")]
+    | OnRule => /* TODO */ [emphasize_text("Rule")]
+    };
+  };
+  switch (typed) {
+  | OnNonLetLine => message(typed) /* Don't display the term tag for empty and comment lines */
+  | _ => List.cons(term_tag, message(typed))
+  };
+};
+
+let novice_summary =
+    (
```

Comment:
yeah, we have the duplicated logic from advanced_summary in the three cases mentioned above... easy to get out of sync. if we could push that up the chain, these would both be clean wrappers on core